utils/format_doc.mli

```diff
@@ -0,0 +1,290 @@
+  | Close_tag
+  | Open_tbox
+  | Tab_break of { width : int; offset : int }
+  | Set_tab
+  | Close_tbox
+  | Simple_break of { spaces : int; indent : int }
+  | Break of { fits : string * int * string as 'a; breaks : 'a }
+  | Flush of { newline:bool }
+  | Newline
+  | If_newline
+
+  | Deprecated of (Format.formatter -> unit)
+    (** Escape hatch: a {!Format} printer used to provide backward-compatibility
+        for user-defined printer (from the [#install_printer] toplevel directive
+        for instance). *)
+
+ (** Immutable document type*)
+type t
+type doc = t
+
+(** Empty document *)
+val empty: t
+
+(** [format ppf doc] sends the format instruction of [doc] to the Format's
+    formatter [doc]. *)
+val format: Format.formatter -> doc -> unit
+
+(** Fold over a document as a sequence of instructions *)
+val fold: ('acc -> element -> 'acc) -> 'acc -> doc -> 'acc
+
+(** Immutable API for composing {!doc} *)
+module Core: sig
+  type ('a,'b) fmt = ('a, doc, doc,'b) format4
+
+  type printer0 = doc -> doc
+  type 'a printer = 'a -> printer0
+
+
+  (** {!msg} and {!kmsg} produce a document from a format string and its
+      argument *)
+  val msg: ('a,doc) fmt -> 'a
+  val kmsg: (doc -> 'b) -> ('a,'b) fmt -> 'a
+
+  (** {!printf} and {!kprintf} produce a printer from a format string and its
+      argument*)
+  val printf: ('a, printer0) fmt -> 'a
+  val kprintf: (doc -> 'b) -> ('a, doc -> 'b) fmt -> 'a
+
+  (** The functions below mirror {!Format} printers, without the [pp_print_]
+      prefix naming convention *)
+  val open_box: box_type -> int -> printer0
+  val close_box: printer0
+
+  val text: string printer
+  val string: string printer
+  val bytes: bytes printer
+  val with_size: int printer
+
+  val int: int printer
+  val float: float printer
+  val char: char printer
+  val bool: bool printer
+
+  val space: printer0
+  val cut: printer0
+  val break: spaces:int -> indent:int -> printer0
+
+  val custom_break:
+    fits:(string * int * string as 'a) -> breaks:'a -> printer0
+  val force_newline: printer0
+  val if_newline: printer0
+
+  val flush: printer0
+  val force_stop: printer0
+
+  val open_tbox: printer0
+  val set_tab: printer0
+  val tab: printer0
+  val tab_break: width:int -> offset:int -> printer0
+  val close_tbox: printer0
+
+  val open_tag: stag printer
+  val close_tag: printer0
+
+  val list: ?sep:printer0 -> 'a printer -> 'a list printer
+  val iter:
+    ?sep:printer0 -> iter:(('a -> unit) -> 'b -> unit) -> 'a printer
+    ->'b printer
+  val array: ?sep:printer0 -> 'a printer -> 'a array printer
+  val seq: ?sep:printer0 -> 'a printer -> 'a Seq.t printer
+
+  val option: ?none:printer0 -> 'a printer -> 'a option printer
+  val result: ok:'a printer -> error:'e printer -> ('a,'e) result printer
+  val either: left:'a printer -> right:'b printer -> ('a,'b) Either.t printer
+
+end
+
+(** {1 Compatibility API} *)
+
+(** The functions and types below provides source compatibility with format
```

Comment:
Opening this module adds the document types, `fold`, etc.. into the environment. What do you think of moving the compatibility API into a `Compat` submodule ?
Also, highlighting that it's at the same level as the `Core` API.

Reply:
One my design goal was to make the replacement of `Format` by `Format_doc` as invisible as possible.
Typically, the `Core` submodule is only exposed because I though it would be a shame to have a simple immutable API available and hide it away. I am not sure I would encourage people to switch to this API.

From that perspective, I think it is better for now to keep the most used features (the compatibility functions) at the toplevel of the module. But I could hide the core definitions inside the `Core` submodule?

Reply:
I have moved the type definitions and core functions inside the `Core` submodule, and renamed it to `Doc`. Like this, an `open Format_doc` only opens the compatibility API and the `Doc` submodule.